contribs/gnodev/pkg/dev/node_test.go

```diff
@@ -0,0 +1,306 @@
+)
+
+// XXX: We should probably use txtar to test this package
+
+// TestNewEmptyDevNode tests the NewDevNode method with no package
+func TestNewNode_NoPackages(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	logger := log.NewTestingLogger(t)
+
+	// Call NewDevNode with no package should works
+	cfg := DefaultNodeConfig(gnoenv.RootDir())
+	node, err := NewDevNode(ctx, logger, &emitter.NoopServer{}, cfg)
+	require.NoError(t, err)
+
+	assert.Len(t, node.ListPkgs(), 0)
+
+	require.NoError(t, node.Close())
+}
+
+// TestNodeWithPackage tests the NewDevNode with a single package.
+func TestNewNode_WithPackage(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const (
+		// foobar package
+		testGnoMod = "module gno.land/r/dev/foobar\n"
+		testFile   = `package foobar
+func Render(_ string) string { return "foo" }
+`
+	)
+
+	// Generate package
+	pkgpath := generateTestingPackage(t, "gno.mod", testGnoMod, "foobar.gno", testFile)
+	logger := log.NewTestingLogger(t)
+
+	// Call NewDevNode with no package should works
+	cfg := DefaultNodeConfig(gnoenv.RootDir())
+	cfg.PackagesPathList = []string{pkgpath}
+	node, err := NewDevNode(ctx, logger, &emitter.NoopServer{}, cfg)
+	require.NoError(t, err)
+	assert.Len(t, node.ListPkgs(), 1)
+
+	// Test rendering
+	render, err := testingRenderRealm(t, node, "gno.land/r/dev/foobar")
+	require.NoError(t, err)
+	assert.Equal(t, render, "foo")
+
+	require.NoError(t, node.Close())
+}
+
+func TestNodeAddPackage(t *testing.T) {
+	// Setup a Node instance
+	const (
+		// foo package
+		fooGnoMod = "module gno.land/r/dev/foo\n"
+		fooFile   = `package foo
+func Render(_ string) string { return "foo" }
+`
+		// bar package
+		barGnoMod = "module gno.land/r/dev/bar\n"
+		barFile   = `package bar
+func Render(_ string) string { return "bar" }
+`
+	)
+
+	// Generate package foo
+	foopkg := generateTestingPackage(t, "gno.mod", fooGnoMod, "foo.gno", fooFile)
+
+	// Call NewDevNode with no package should works
+	node, emitter := newTestingDevNode(t, foopkg)
+	assert.Len(t, node.ListPkgs(), 1)
+
+	// Test render
+	render, err := testingRenderRealm(t, node, "gno.land/r/dev/foo")
+	require.NoError(t, err)
+	require.Equal(t, render, "foo")
+
+	// Generate package bar
+	barpkg := generateTestingPackage(t, "gno.mod", barGnoMod, "bar.gno", barFile)
+	err = node.UpdatePackages(barpkg)
+	require.NoError(t, err)
+	assert.Len(t, node.ListPkgs(), 2)
+
+	// Render should failed as the node havn't reload
+	render, err = testingRenderRealm(t, node, "gno.land/r/dev/bar")
+	require.Error(t, err)
+
+	err = node.Reload(context.Background())
+	require.NoError(t, err)
+	assert.Equal(t, emitter.NextEvent().Type(), events.EvtReload)
+
+	// After a reload, render should succeed
+	render, err = testingRenderRealm(t, node, "gno.land/r/dev/bar")
+	require.NoError(t, err)
+	require.Equal(t, render, "bar")
+}
+
+func TestNodeUpdatePackage(t *testing.T) {
```

Comment:
You can run this test in parallel

Reply:
Unfortunately, we should not run Node integration tests in parallel as long as the VM uses global variables